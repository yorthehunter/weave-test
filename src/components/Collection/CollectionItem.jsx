import React from 'react';
import { string, oneOf, oneOfType, bool, node } from 'prop-types';
import CustomTag from '../utilities/CustomTag';
import { sizeModifier } from '../utilities/util';

const componentPrefix = 'weave-collection__item';

/**
 * A collection item is a custom child of a <Collection> that can accept an
 * "inset" which is a padding applied to all <CollectionItem>s and can also
 * produce a UI indicator, typically a color representation of a status.
 */

const CollectionItem = (props) => {
  const {
    inset,
    indicator,
    children,
    isSelected,
  } = props;

  const componentClasses = [
    componentPrefix,
    sizeModifier(inset, 'u-inset'),
    indicator ? `weave-collection-indicator weave-collection-indicator--${indicator}` : '',
  ];

  return (
    <CustomTag componentClasses={componentClasses} {...props}>
      {children}
    </CustomTag>
  );
};

const defaultProps = {
  inset: undefined,
  tag: 'li',
  indicator: undefined,
  isSelected: false,
};

const propTypes = {
  inset: oneOf([undefined, 'sm', 'md', 'lg']),
  tag: string,
  indicator: oneOfType([oneOf(['neutral', 'accent', 'success', 'alert', 'info', 'warning']), bool]),
  isSelected: bool,
  children: node.isRequired,
};

CollectionItem.propTypes = propTypes;
CollectionItem.defaultProps = defaultProps;

export default CollectionItem;
