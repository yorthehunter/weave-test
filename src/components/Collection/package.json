{
  "name": "weave-react-collection",
  "dependencies": {
    "weave-react-utilities": "file:../utilities",
    "weave-ui-collection": "file:../../weave/components/collection"
  },
  "version": "1.0.0",
  "description": "Weave collection React component",
  "main": "index.jsx",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/stitchfix/weave.git"
  },
  "keywords": [
    "css",
    "scss",
    "weave",
    "design",
    "system",
    "style",
    "guide",
    "component",
    "library",
    "ui"
  ],
  "author": "Stitch Fix",
  "license": "UNLICENSED",
  "bugs": {
    "url": "https://github.com/stitchfix/weave/issues"
  },
  "homepage": "https://github.com/stitchfix/weave#readme"
}
