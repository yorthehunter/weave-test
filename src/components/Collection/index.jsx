import React from 'react';
import { string, oneOf, node, bool } from 'prop-types';
import CollectionItem from './CollectionItem';
import CustomTag from '../utilities/CustomTag';
import { sizeModifier } from '../utilities/util';

import '../../weave/components/collection/index.scss';

const componentPrefix = 'weave-collection';

/**
 * A collection is a versatile list of similar items that visually stands apart
 * from other elements on the page.
 */

const Collection = (props) => {
  const {
    inset,
    isHoverable,
    size,
    children,
  } = props;

  const componentClasses = [
    componentPrefix,
    isHoverable ? `${componentPrefix}--hoverable` : null,
    sizeModifier(size, componentPrefix),
  ];

  const renderChildren =
    React.Children.map(children, child => {
      if (child.type === CollectionItem) {
        // Checks children for <CollectionItem> component and passes inset prop
        return React.cloneElement(child, { inset });
      }
      // Will return any other children as normal
      return child;
    });

  return (
    <CustomTag componentClasses={componentClasses} {...props}>
      {renderChildren}
    </CustomTag>
  );
};

const defaultProps = {
  inset: undefined,
  tag: 'ul',
  isHoverable: false,
  size: null,
};

const propTypes = {
  inset: oneOf([undefined, 'sm', 'md', 'lg']),
  tag: string,
  isHoverable: bool,
  size: string,
  children: node.isRequired,
};

Collection.propTypes = propTypes;
Collection.defaultProps = defaultProps;

export default Collection;
