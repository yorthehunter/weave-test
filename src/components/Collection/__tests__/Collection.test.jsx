import React from 'react';
import { render } from 'enzyme';
import Collection from '../';
import CollectionItem from '../CollectionItem';

describe('Collection', () => {
  const props = { inset: 'sm' };
  const wrapper = (
    <Collection {...props}>
      <CollectionItem>Item</CollectionItem>
      <CollectionItem indicator="alert">Item 2</CollectionItem>
      <CollectionItem isSelected>Item 2</CollectionItem>
      <li>Some other child</li>
    </Collection>
  );

  it('renders correctly with default values', () => {
    const rendered = render(wrapper);
    expect(rendered).toMatchSnapshot();
  });
});
