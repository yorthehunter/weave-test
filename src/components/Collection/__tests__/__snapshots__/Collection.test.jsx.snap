// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection renders correctly with default values 1`] = `
<ul
  class="weave-collection"
>
  <li
    class="weave-collection__item u-inset--sm"
    style="background-color:transparent"
  >
    Item
  </li>
  <li
    class="weave-collection__item u-inset--sm weave-collection-indicator weave-collection-indicator--alert"
    style="background-color:transparent"
  >
    Item 2
  </li>
  <li
    class="weave-collection__item u-inset--sm"
    style="background-color:#D9F5F4"
  >
    Item 2
  </li>
  <li>
    Some other child
  </li>
</ul>
`;
